crates/sats/src/slim_slice.rs

```diff
@@ -0,0 +1,1105 @@
+    #[allow(clippy::needless_lifetimes)]
+    unsafe fn deref_mut<'a>(&'a mut self) -> &'a mut [T] {
+        let (ptr, len) = self.split();
+        // SAFETY: caller is responsible for these.
+        unsafe { slice::from_raw_parts_mut(ptr, len) }
+    }
+
+    /// Creates the raw slice from a pointer to the data and a length.
+    ///
+    /// It is assumed that `len <= u32::MAX`.
+    /// The caller must ensure that `ptr != NULL`.
+    const fn from_len_ptr(len: usize, ptr: *mut T) -> Self {
+        // SAFETY: caller ensured that `!ptr.is_null()`.
+        let ptr = unsafe { NonNull::new_unchecked(ptr) };
+        let len = len as u32;
+        Self { ptr, len }
+    }
+}
+
+impl<T> Copy for SlimRawSlice<T> {}
+impl<T> Clone for SlimRawSlice<T> {
+    fn clone(&self) -> Self {
+        *self
+    }
+}
+
+// =============================================================================
+// Owned boxed slice
+// =============================================================================
+
+/// Provides a slimmer version of `Box<[T]>`
+/// using `u32` for its length instead of `usize`.
+#[repr(transparent)]
+pub struct SlimSliceBox<T> {
+    /// The representation of this boxed slice.
+    raw: SlimRawSlice<T>,
+    /// Marker to ensure covariance and dropck ownership.
+    owned: PhantomData<T>,
+}
+
+/// `SlimSliceBox<T>` is `Send` if `T` is `Send` because the data is owned.
+unsafe impl<T: Send> Send for SlimSliceBox<T> {}
+
+/// `SlimSliceBox<T>` is `Sync` if `T` is `Sync` because the data is owned.
+unsafe impl<T: Sync> Sync for SlimSliceBox<T> {}
+
+impl<T> Drop for SlimSliceBox<T> {
+    fn drop(&mut self) {
+        // Get us an owned `SlimSliceBox<T>`
+        // by replacing `self` with garbage that won't be dropped
+        // as the drop glue for the constituent fields does nothing.
+        let ptr = NonNull::dangling();
+        let raw = SlimRawSlice { len: 0, ptr };
+        let owned = PhantomData;
+        let this = mem::replace(self, Self { raw, owned });
+
+        // Convert into `Box<[T]>` and let it deal with dropping.
+        drop(into_box(this));
+    }
+}
+
+impl<T> SlimSliceBox<T> {
+    /// Converts `boxed` to `Self` without checking `boxed.len() <= u32::MAX`.
+    ///
+    /// Only safe to call if the constraint above is satisfied.
+    #[allow(clippy::boxed_local)]
+    // Clippy doesn't seem to consider unsafe code here.
+    unsafe fn from_boxed_unchecked(boxed: Box<[T]>) -> Self {
+        let len = boxed.len();
+        let ptr = Box::into_raw(boxed) as *mut T;
+        // SAFETY: `Box<T>`'s ptr was a `NonNull<T>` already.
+        // and our caller has promised that `boxed.len() <= u32::MAX`.
+        let raw = SlimRawSlice::from_len_ptr(len, ptr);
+        let owned = PhantomData;
+        Self { raw, owned }
+    }
+
+    /// Converts `boxed: Box<[T]>` into `SlimSliceBox<T>`.
+    ///
+    /// Panics when `boxed.len() > u32::MAX`.
+    pub fn from_boxed(boxed: Box<[T]>) -> Self {
```

Comment:
I think I'd like the comment about "all paths to Self go through Box<[T]>" here where the creation actually happens, as well.